src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -55,57 +61,196 @@ internal static partial class JsonWriterHelper
+
+                        // Check if any of the 16 bytes need to be escaped.
+                        Vector128<sbyte> mask = CreateEscapingMask(sourceValue);
+
+                        int index = Sse2.MoveMask(mask.AsByte());
+                        // If index == 0, that means none of the 16 bytes needed to be escaped.
+                        // TrailingZeroCount is relatively expensive, avoid it if possible.
+                        if (index != 0)
+                        {
+                            // Found at least one byte that needs to be escaped, figure out the index of
+                            // the first one found that needed to be escaped within the 16 bytes.
+                            idx += BitOperations.TrailingZeroCount(index | 0xFFFF0000);
+                            goto Return;
+                        }
+                        idx += 16;
+                        startingAddress += 16;
+                    }
+
+                    // Process the remaining characters.
+                    Debug.Assert(value.Length - idx < 16);
+                }
+#endif
+
+                for (; idx < value.Length; idx++)
+                {
+                    Debug.Assert((ptr + idx) <= (ptr + value.Length));
+                    if (NeedsEscaping(*(ptr + idx)))
+                    {
+                        goto Return;
+                    }
+                }
 
-        Return:
-            return idx;
+                idx = -1; // all characters allowed
+
+            Return:
+                return idx;
+            }
         }
 
         public static unsafe int NeedsEscaping(ReadOnlySpan<char> value, JavaScriptEncoder encoder)
         {
-            int idx;
-
-            // Some implementations of JavascriptEncoder.FindFirstCharacterToEncode may not accept
-            // null pointers and gaurd against that. Hence, check up-front and fall down to return -1.
-            if (encoder != null && !value.IsEmpty)
+            fixed (char* ptr = value)
             {
-                fixed (char* ptr = value)
+                int idx = 0;
+
+                // Some implementations of JavascriptEncoder.FindFirstCharacterToEncode may not accept
+                // null pointers and gaurd against that. Hence, check up-front and fall down to return -1.
+                if (encoder != null && !value.IsEmpty)
                 {
                     idx = encoder.FindFirstCharacterToEncode(ptr, value.Length);
+                    goto Return;
                 }
-                goto Return;
-            }
 
-            for (idx = 0; idx < value.Length; idx++)
-            {
-                if (NeedsEscaping(value[idx]))
+#if BUILDING_INBOX_LIBRARY
+                if (Sse2.IsSupported)
                 {
-                    goto Return;
+                    short* startingAddress = (short*)ptr;
+                    while (value.Length - 8 >= idx)
```

Comment:
is `value.Length - 8` cached or recomputed each loop iteration?

Reply:
It's effectively cached. The generated assembly doesn't change if I save it in a local.
```text
00007FFD786C17CA  mov         rbp,rdi  
00007FFD786C17CD  lea         r14d,[rsi-8]  
00007FFD786C17D1  test        r14d,r14d  
00007FFD786C17D4  jl          00007FFD786C180A  
```